Decryption share verification, updates

iquerejeta opened the pull request:
Continued with the work of `decrypt_nizk`, which was initially opened at #542 (this PR supersedes it): 
* resolved bug in fiat shamir
* resolved design bug of the ZKP
* and some docs and rebase with latest master

Zeegomo commented on chain-vote/src/decr_nizk.rs:
```diff
@@ -0,0 +1,146 @@
+
+/// Proof of correct decryption.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct ProofDecrypt {
```
I was wondering if we should have a common interface for ZK proofs, maybe it's too much overhead. A first step could be grouping them under a common directory / module

iquerejeta replied:
Yes, I agree. Maybe we can do that refactor in a different PR?

Zeegomo commented on chain-vote/src/decr_nizk.rs:
```diff
@@ -0,0 +1,146 @@
+
+    pub fn to_bytes(&self) -> [u8; PROOF_SIZE] {
+        let mut output = [0u8; PROOF_SIZE];
+        output[0..GroupElement::BYTES_LEN].copy_from_slice(&self.a1.to_bytes());
```
This can reuse `to_slice_mut`

Zeegomo commented on chain-vote/src/decr_nizk.rs:
```diff
@@ -0,0 +1,146 @@
+    use crate::encryption::Keypair;
+    use rand_chacha::ChaCha20Rng;
+    use rand_core::SeedableRng;
+    // use chain_crypto::algorithms::vrf::dleq::*;
```
leftover?

iquerejeta commented on chain-vote/src/lib.rs:
```diff
@@ -242,12 +290,25 @@ fn group_elements_from_bytes(bytes: &[u8]) -> Option<Vec<gang::GroupElement>> {
     Some(elements)
 }
 
+pub fn verify_decrypt_share(
```
yes, good point. I was thinking of changing the `lib.rs` file, and try to keep consistency with what we have in the rest of the library (basically, just defining the modules). I initially left it out of this PR because the goal was to compare against `decrypt_nizk`, but given that we are now doing the PR against `master`, do you think it makes sense to do this refactor here?

Zeegomo replied:
Yep, it does. Just lay out whatever you think is the best structure

iquerejeta replied:
Ok, some major refactor coming then

iquerejeta replied:
agreed to put it in a different PR

Zeegomo commented on chain-vote/src/lib.rs:
```diff
@@ -382,9 +460,17 @@ mod tests {
+        let (ts, tds3) = tally.finish(&mut rng, m3.secret_key());
+
+        // check that the verify shares are correct for each participants
+        assert_eq!(verify_decrypt_share(&tally, &m1.public_key(), &tds1), true);
```
Is this check redundant now that shares are verified with the tally?

iquerejeta replied:
yep, just kept one for checking that the function fails with incorrect key

iquerejeta commented:
Major refactor included: 
* new cryptography module with zkps, commitment and elgamal
* zkps module contains decryption and unit_vector zkps
* moved code in `lib.rs` to its own module, under `tally.rs`
   * all functions are now associated functions to a given structure
   * removed the `TallyState` structure, which was redundant
   * added documentation
* Deleted `EncryptingVote` structure
* made `encrypt_and_prove_vote` as an associated function to
  `ElectionPublicKey`
* Generalised UnitVectorZkp to be abstracted from election notation
* Added intermediary struct ValidatedTally, to enforce that decryption only happens given valid shares (this will be helpful in the future to handle threshold decryption)
* Enforced that only valid votes can be added to the encrypted tally (this still fails in the mockchain-repo---requires a few more changes)

iquerejeta commented on chain-vote/src/lib.rs:
```diff
@@ -271,12 +332,27 @@ pub fn tally(
+        pks: &[committee::MemberPublicKey],
+        tally_state: &TallyState,
+        decrypt_shares: &[TallyDecryptShare],
+    ) -> bool {
```
Yes, I agree. But I'm working on custom errors in the refactor. Probably best to change it in the next PR 👍

Zeegomo replied:
Ok, good to go then, good job!
